sharedbuffers/mapped_struct.py

```diff
@@ -2817,12 +2857,117 @@ def unpack_from(cls, buf, offs, idmap = None):
+
+    Timestamps are stored in UTC. If the datetime object does have a tzinfo the
+    conversion is guaranteed to be correct. If not, it will assume it's local time
+    or what's set in NAIVE_TZ, and if not the conversion will not be correct.
+
+    If it is expected that naive datetime objects with non-local times will be used,
+    use a subclass of this setting the expected TZ in the NAIVE_TZ class attribute.
+
+    Unpacked values will always be datetime objects with the same timezone of the original
+    value, so it preserves tz info, or NAIVE_TZ if it was given a naive datetime.
+    """
+    PACKER = struct.Struct('=qh')
+
+    NAIVE_TZ = TZUTC
+    ID_MASK = DT_TZ_MASK
+
+    _TZ_CACHE = {}
+
+    @classmethod
+    @cython.locals(offs = cython.longlong, implicit_offs = cython.longlong, timestamp = cython.longlong,
+        widmap = StrongIdMap)
+    def pack_into(cls, obj, buf, offs, idmap = None, implicit_offs = 0):
+        if idmap is not None:
+            objid = shared_id(obj)
+            id_mask = cls.ID_MASK
+            if id_mask is not None:
+                objid |= id_mask
+            idmap[objid] = offs + implicit_offs
+            if isinstance(idmap, StrongIdMap):
+                widmap = idmap
+                widmap.link(objid, obj)
+
+        if obj.tzinfo is None:
+            naive_tz = cls.NAIVE_TZ
+            if naive_tz is not None and callable(naive_tz):
+                naive_tz = naive_tz()
+            if naive_tz is not None:
+                obj = obj.replace(tzinfo=naive_tz)
+
+        packer = cls.PACKER
+        timestamp = int(calendar.timegm(obj.astimezone(TZUTC).timetuple()))
+        utc_off = int(obj.utcoffset().total_seconds() / 60)
+        packer.pack_into(buf, offs, (timestamp << 20) + obj.microsecond, utc_off)
+
+        return offs + packer.size
+
+    @classmethod
+    @cython.locals(offs = cython.longlong, timestamp = cython.longlong, microseconds = cython.ulong)
+    def unpack_from(cls, buf, offs, idmap = None):
+        if idmap is not None and offs in idmap:
+            return idmap[offs]
+
+        tz_cache = cls._TZ_CACHE
+        packer = cls.PACKER
+        timestamp, utc_off = packer.unpack_from(buf, offs)
+        microseconds = timestamp & 0xFFFFF
+        tz_info = tz_cache.get(utc_off)
+        if tz_info is None:
+            if utc_off:
+                tz_name = 'UTC%+d' % (utc_off / 60,)
+                if utc_off % 60:
+                    tz_name = "%s:%02d" % (tz_name, utc_off % 60)
+                tz_info = dateutil.tz.tzoffset(tz_name, utc_off * 60)
+            else:
+                tz_info = TZUTC
+            tz_info = tz_cache.setdefault(utc_off, tz_info)
+        rv = datetime.fromtimestamp(timestamp >> 20, tz_info) + timedelta(microseconds=microseconds)
+
+        if idmap is not None:
+            idmap[offs] = rv
+        return rv
+
+
+class mapped_datetime_tz_local(datetime):
```

Comment:
Is it the parent class ok? Don't have to be mapped_datetime_tz?

Reply:
No, it's a complete implementation so I made it separate.

Reply:
Wait, this is the tz_local, yes